### PR TITLE
[Fiber] "Task" priority for error boundaries and batched updates

### PR DESCRIPTION
**Pull request by acdlite**
Introduces a new priority level, **Task**. (The name is provisional. We'll come up with something better later.)

Task priority is similar to Synchronous priority. Both are flushed in the current tick. The difference is

- work with the **Task** priority is performed within the current tick, but at the end of the batch of work (after all work has been committed).
- work with the **Synchronous** priority is performed immediately. E.g. a synchronous `setState` will perform all its work before the `setState` call exits.

In https://github.com/facebook/react/pull/8127, these two types of work priority were both considered "synchronous." Splitting them into separate concepts will make it easier to schedule and batch things correctly.

Currently used for `batchedUpdates` and nested sync updates. Task should also be used for updates inside `componentDidUpdate/Mount`, error boundary work, and events. I'll leave these for a future PR.

There's an error boundary regression that I haven't looked at yet. I'll investigate it in the morning. Pushing this now because I'm about to go to bed and I don't want to block @gaearon.

### CODE REVIEW

**Comment by acdlite** [comment]
Pushed a new commit that passes all the error boundary tests. Changed the algorithm of `handleErrors` a bit to ensure that boundaries are not revisited once they are acknowledged, which prevents the same boundary from being visited over and over again. I had to move the functions in `ReactFiberErrorBoundary.js` into the scheduler so they have access to its state.

This passes all the error boundary tests and the incremental tests; however, it's causing an infinite loop in one of the files when I try to run the full suite. Haven't figured out which one yet. Unfortunately, the infinite loop means I can't run the `record-tests` script to see which tests might be failing. Working on Fiber is like whack-a-mole sometimes :D

I feel pretty good about the overall approach here, though. Will continue to investigate which test is breaking tonight; if I can't find it, I'll pick it up in the morning.

@gaearon

**Comment by acdlite** [comment]
Found it! The test is `should only call componentWillUnmount once` in `ReactCompositeComponent-test.js`. Makes sense.

**Comment by acdlite** [comment]
Yay! I got it working!

Could use a bit more cleanup, but this should be ready for review @gaearon.

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 1]
```diff
@@ -594,98 +522,203 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
...
     }
   }
 
+  function findClosestErrorBoundary(fiber : Fiber): Fiber | null {
+    let maybeErrorBoundary = fiber.return;
+    while (maybeErrorBoundary) {
+      if (maybeErrorBoundary.tag === ClassComponent) {
+        const instance = maybeErrorBoundary.stateNode;
+        if (typeof instance.unstable_handleError === 'function' &&
+            !knownBoundaries.has(maybeErrorBoundary)) {
```
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: This line here is how I got it to work, and why I had to move this function into the scheduler.

File: src/renderers/dom/shared/ReactDOMFeatureFlags.js [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 var ReactDOMFeatureFlags = {
   useCreateElement: true,
-  useFiber: false,
+  useFiber: true,
```
Comment (gaearon) [inline src/renderers/dom/shared/ReactDOMFeatureFlags.js]: Careful

Comment (gaearon) [inline src/renderers/dom/shared/ReactDOMFeatureFlags.js]: Unless you consider it ready for 15.4.0 😄

Comment (acdlite) [inline src/renderers/dom/shared/ReactDOMFeatureFlags.js]: Hehe :D

**Pull request merged**